src/hotspot/share/gc/shared/pretouchTask.cpp

```diff
@@ -28,59 +28,99 @@
 #include "runtime/atomic.hpp"
 #include "runtime/globals.hpp"
 #include "runtime/os.hpp"
+#include "utilities/ticks.hpp"
 
 PretouchTask::PretouchTask(const char* task_name,
                            char* start_address,
                            char* end_address,
                            size_t page_size,
-                           size_t chunk_size) :
+                           size_t chunk_size,
+                           uint n_threads,
+                           size_t task_status) :
     AbstractGangTask(task_name),
     _cur_addr(start_address),
     _start_addr(start_address),
     _end_addr(end_address),
     _page_size(page_size),
-    _chunk_size(chunk_size) {
+    _chunk_size(chunk_size),
+    _n_threads(n_threads),
+    _task_status(task_status) {
 
   assert(chunk_size >= page_size,
          "Chunk size " SIZE_FORMAT " is smaller than page size " SIZE_FORMAT,
          chunk_size, page_size);
 }
 
+void PretouchTask::reinitialize(char* start_addr, char* end_addr) {
+  Atomic::release_store(&_cur_addr, start_addr);
+  Atomic::release_store(&_end_addr, end_addr);
+}
+
 size_t PretouchTask::chunk_size() {
   return PreTouchParallelChunkSize;
 }
 
 void PretouchTask::work(uint worker_id) {
+
+  // Following atomic loads are required to make other processor store
+  // visible to all threads from this points.
```

Comment:
That is not what an Atomic::load does; they only protect against word-tearing (which is only a theoretical possibility on some platforms). If you want to guarantee those loads see the most recent store then the fence needs to come first.

Reply:
Thank you David for your reply and will do as suggested.